new plugin - one-move-per-press

jon-heard opened the pull request.

jon-heard committed:
new plugin - one-move-per-press

seleb commented on src/one move per press.js:
although this seems to work as intended, i wanted to raise a concern with the approach here:

this plugin is addressing an accessibility issue, but it being a plugin limits the practical use quite a bit. it requires a game's developer to decide to include it, and including it also _reduces_ accessibility for other players since the control scheme is modified for everyone regardless of their needs. for your use case (experienced dev helping children make a game, probably not distributing widely) this plugin works fine, but the root problem here ("bipsi's movement cooldown may be too sensitive for players with poor motor control") should probably be discussed with candle. unfortunately this isn't a trivial thing to fix since bipsi does not currently have a settings menu or etc where something like the delay could be customized.

for reference, the OS does allow users to customize the delay and rate of input repeats for this reason, e.g. on windows:
![windows keyboard properties showing character repeat delay/rate](https://github.com/seleb/bipsi-hacks/assets/6496840/40c3d3f9-3f47-4d8f-9493-eee50a9e6529)

it's possible bipsi's input handling could be refactored in such a way that these delays are inherited, but i believe it is an intentional design choice to avoid this since the initial delay is typically much longer than the repeat, which can feel clunky in a game

some other minor notes:
- this only modifies the keyboard input and leaves the touch input as-is, which could be confusing
- the way this is currently implemented is to intercept the same keys bipsi expects and re-dispatch them conditionally. this means you've necessarily duplicated the bipsi key handling, but it may be simpler + more future-proof if you prevented player movement via a method that isn't specific to the input (e.g. set `busy` when any key is held, modify the cooldown delay directly, etc)

jon-heard replied:
I was on the fence about submitting this for the reasons you've raised: it's such a niche plugin.  However, I figured that if someone needs it then they probably *really* need it.

> it may be simpler + more future-proof if you prevented player movement via a method that isn't specific to the input

Unfortunately, the input system is fairly well encapsulated within bipsi, including movement keys and cooldown delay.  I'll look into setting the `busy` flag when any key is held, though.

jon-heard committed:
updates to one move per press:
- can toggle the plugin through code, allowing for game settings to set this feature
- not tied to particular movement keys

jon-heard committed:
one move per press - Added the limiter to touch movement as well.

jon-heard committed:
one move per press - fixed documentation wrapping

jon-heard commented:
I've updated the plugin.
- It no longer relies on the specific keys being pressed
- The plugin can be disabled to allow for a game's settings to control this feature
- Impacts touch as well as keyboard and gamepad

seleb committed:
Merge branch 'main' into newPlugin_oneMovePerPress

seleb commented:
:tada: This PR is included in version 8.1.0 :tada:

The release is available on [GitHub release](https://github.com/seleb/bipsi-hacks/releases/tag/v8.1.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
